edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs

```diff
@@ -0,0 +1,338 @@
+            //  - The testcode generates clients (both module and device, and direct and nested), all clients subscribe to
+            //    desired property change, so after reconnection, edgeHub should re-subscribe for all clients.
+            var (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector) = await SetupEnvironment();
+            var allClients = await ConnectClientsAsync(subscriptionChangeHandler);
+
+            await Task.Delay(2000);
+            await cloudProxyDispatcher.HandleAsync(new MqttPublishInfo("$internal/connectivity", Encoding.UTF8.GetBytes("{\"status\":\"Connected\"}")));
+
+            await Task.Delay(2000);
+            await cloudProxyDispatcher.HandleAsync(new MqttPublishInfo("$internal/connectivity", Encoding.UTF8.GetBytes("{\"status\":\"Disconnected\"}")));
+
+            await Task.Delay(2000);
+
+            var clients = new HashSet<IIdentity>();
+            brokerConnector.SetPacketSpy(Spy);
+
+            await cloudProxyDispatcher.HandleAsync(new MqttPublishInfo("$internal/connectivity", Encoding.UTF8.GetBytes("{\"status\":\"Connected\"}")));
+
+            // We don't know when all the clients get processed, so keep looping for a while
+            var startTime = DateTime.Now;
+            while (DateTime.Now - startTime < TimeSpan.FromSeconds(30))
+            {
+                bool isCountEqual;
+                lock (clients)
+                {
+                    isCountEqual = allClients.Count == clients.Count;
+                }
+
+                if (isCountEqual)
+                {
+                    break;
+                }
+                else
+                {
+                    await Task.Delay(100);
+                }
+            }
+
+            Assert.True(allClients.SetEquals(clients));
+
+            void Spy(RpcPacket packet)
+            {
+                if (packet.Topic.EndsWith("/methods/post/#"))
+                {
+                    var segments = packet.Topic.Split('/');
+
+                    lock (clients)
+                    {
+                        if (segments.Length == 6)
+                        {
+                            clients.Add(new ModuleIdentity(iotHubName, segments[1], segments[2]));
+                        }
+                        else if (segments.Length == 5)
+                        {
+                            clients.Add(new DeviceIdentity(iotHubName, segments[1]));
+                        }
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public async Task OnReconnectionClientsGetTwinsPulled()
+        {
+            var (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector) = await SetupEnvironment();
+            var allClients = await ConnectClientsAsync(subscriptionChangeHandler);
+
+            await cloudProxyDispatcher.HandleAsync(new MqttPublishInfo("$internal/connectivity", Encoding.UTF8.GetBytes("{\"status\":\"Connected\"}")));
```

Comment:
I know this is test code, but this looks similar to code above, not sure what our code guidelines are for C#, but could this go in some setup method?

Reply:
I moved some common pieces into methods to reuse